File: estilizar_ruby_streaks.py
import os
from PIL import Image, ImageFilter, ImageEnhance

# 1. Cargar la imagen de origen (convirtiendo a RGBA para transparencia)
source_path = os.path.join(os.path.dirname(__file__), "hoja_ruby_streaks_raw.jpg")
if not os.path.exists(source_path):
    # Buscar en rutas alternativas
    for candidate in ["hoja_ruby_streaks_raw.jpg", "ruby_streaks_leaf.jpg", "assets/productos/mostaza_ruby_streaks.jpg"]:
        if os.path.exists(candidate):
            source_path = candidate
            break

img = Image.open(source_path).convert("RGBA")

# 2. Remover matemáticamente el fondo blanco
datas = img.getdata()
newData = []
for item in datas:
    # Si los canales R, G, B son muy cercanos al blanco, los hacemos transparentes
    if item[0] > 235 and item[1] > 235 and item[2] > 235:
        newData.append((255, 255, 255, 0))
    else:
        newData.append(item)
img.putdata(newData)

# 3. Recortar bordes vacíos sobrantes
bbox = img.getbbox()
if bbox:
    img = img.crop(bbox)

# 4. Crear un lienzo de fondo con proporción 4:3 (600x450 px) de tono gris-cálido premium
bg = Image.new("RGBA", (600, 450), (242, 243, 240, 255))

# 5. Redimensionar la hoja para que tenga una altura elegante de 340px manteniendo proporción
leaf_height = 340
h_percent = (leaf_height / float(img.size[1]))
w_size = int((float(img.size[0]) * float(h_percent)))
leaf_resized = img.resize((w_size, leaf_height), Image.Resampling.LANCZOS)

# 6. Realzar el color (hacer que el tono rubí/púrpura sea vibrante y fresco)
color_enhancer = ImageEnhance.Color(leaf_resized)
leaf_colored = color_enhancer.enhance(1.4)  # Boost de color del 40%
contrast_enhancer = ImageEnhance.Contrast(leaf_colored)
leaf_final = contrast_enhancer.enhance(1.15) # Ligero boost de contraste

# 7. Generar una sombra de gota (Drop Shadow) difuminada para efecto 3D
shadow = Image.new("RGBA", leaf_final.size, (0, 0, 0, 0))
shadow_data = []
for item in leaf_final.getdata():
    if item[3] > 0:
        # Sombra suave de tono carbón con opacidad ligera
        shadow_data.append((45, 49, 46, int(item[3] * 0.15)))
    else:
        shadow_data.append((0, 0, 0, 0))
shadow.putdata(shadow_data)
shadow_blurred = shadow.filter(ImageFilter.GaussianBlur(12)) # Desenfoque suave

# 8. Centrar y pegar los elementos en el lienzo
x = (600 - w_size) // 2
y = (450 - leaf_height) // 2

# Pegar sombra con un sutil desplazamiento (offset) hacia la derecha y abajo
bg.paste(shadow_blurred, (x + 8, y + 8), shadow_blurred)
# Pegar la hoja estilizada
bg.paste(leaf_final, (x, y), leaf_final)

# 9. Guardar en la carpeta local de productos en formato WebP de alta velocidad
os.makedirs("assets/productos", exist_ok=True)
bg.convert("RGB").save("assets/productos/mostaza_ruby_streaks.webp", "WEBP", quality=90)
print("¡Estilización botánica de Ruby Streaks completada con éxito!")

File: estilizar_ruby_streaks.js
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');

async function processLeaf() {
  console.log("Iniciando procesamiento y estilización botánica de Mostaza Ruby Streaks...");
  
  const sourcePath = path.join(__dirname, 'hoja_ruby_streaks_raw.jpg');
  if (!fs.existsSync(sourcePath)) {
    throw new Error("No se encontró hoja_ruby_streaks_raw.jpg");
  }

  // 1. Cargar imagen y obtener datos en crudo (raw RGBA)
  const image = sharp(sourcePath).ensureAlpha();
  const { data, info } = await image.raw().toBuffer({ resolveWithObject: true });
  
  const width = info.width;
  const height = info.height;
  const channels = info.channels;
  
  // 2. Remover fondo blanco matemáticamente (R > 235, G > 235, B > 235 -> Alpha 0)
  for (let i = 0; i < data.length; i += channels) {
    const r = data[i];
    const g = data[i + 1];
    const b = data[i + 2];
    if (r > 235 && g > 235 && b > 235) {
      data[i + 3] = 0; // Transparente
    }
  }

  // 3. Reconstruir PNG con transparencia y recortar bordes vacíos
  const transparentPng = await sharp(data, {
    raw: { width, height, channels: 4 }
  })
  .png()
  .toBuffer();

  const trimmedPng = await sharp(transparentPng)
    .trim()
    .png()
    .toBuffer();

  // 4. Redimensionar hoja a altura elegante de 340px manteniendo proporción
  // 6. Realzar saturación (+40%) y contraste (+15%)
  const styledLeafBuffer = await sharp(trimmedPng)
    .resize({ height: 340, fit: 'inside' })
    .modulate({ saturation: 1.4, brightness: 1.02 })
    .linear(1.15, -(128 * 0.15))
    .png()
    .toBuffer();

  const leafMeta = await sharp(styledLeafBuffer).metadata();
  const leafW = leafMeta.width;
  const leafH = leafMeta.height;

  // 7. Generar Drop Shadow difuminada (tono carbón rgba(45, 49, 46, 0.15) con Gaussian blur)
  const leafRaw = await sharp(styledLeafBuffer).raw().toBuffer({ resolveWithObject: true });
  const shadowData = Buffer.alloc(leafRaw.data.length);
  for (let i = 0; i < leafRaw.data.length; i += 4) {
    const alpha = leafRaw.data[i + 3];
    if (alpha > 0) {
      shadowData[i] = 45;     // R carbón
      shadowData[i + 1] = 49; // G carbón
      shadowData[i + 2] = 46; // B carbón
      shadowData[i + 3] = Math.round(alpha * 0.18); // Opacidad de sombra
    } else {
      shadowData[i + 3] = 0;
    }
  }

  const shadowBuffer = await sharp(shadowData, {
    raw: { width: leafW, height: leafH, channels: 4 }
  })
  .png()
  .toBuffer();

  const shadowBlurred = await sharp(shadowBuffer)
    .blur(12)
    .png()
    .toBuffer();

  // 8. Crear lienzo 600x450 px tono gris-cálido premium (#F2F3F0)
  const canvasW = 600;
  const canvasH = 450;
  const posX = Math.round((canvasW - leafW) / 2);
  const posY = Math.round((canvasH - leafH) / 2);

  const background = sharp({
    create: {
      width: canvasW,
      height: canvasH,
      channels: 4,
      background: { r: 242, g: 243, b: 240, alpha: 1 }
    }
  });

  // 9. Componer sombra con offset (+8, +8) y hoja centrada
  const outPath = path.join(__dirname, 'assets', 'productos', 'mostaza_ruby_streaks.webp');
  
  await background
    .composite([
      { input: shadowBlurred, top: Math.min(canvasH - leafH, posY + 8), left: Math.min(canvasW - leafW, posX + 8) },
      { input: styledLeafBuffer, top: posY, left: posX }
    ])
    .webp({ quality: 90, effort: 6 })
    .toFile(outPath);

  const stats = fs.statSync(outPath);
  console.log(`¡Estilización botánica de Ruby Streaks completada con éxito!`);
  console.log(`Guardado en: ${outPath} (${Math.round(stats.size / 1024)} KB)`);
}

processLeaf().catch(console.error);
